update GD32F30x firmware library to 2.1.4

tlyu opened the pull request:
Vendor release 2.1.4 dated 2022-07-11. URLs:
https://www.gd32mcu.com/download/down/document_id/129/path_type/1
https://www.gd32mcu.com/data/documents/toolSoftware/GD32F30x_Firmware_Library_V2.1.4.rar

Notable USBD library improvements:

* Track control transfer state, avoiding some race conditions that
  caused looping ping-pong status transactions.

* Change some API types from 8-bit to 16-bit, avoiding truncation.

* Handle configuration descriptor sets longer than 255 bytes.

This PR also re-applies some local patches, and undoes an incorrect vendor change to the USBD library.

tlyu committed:
re-apply I2C clock change patch

Signed-off-by: Taylor Yu <code@argon.blue>

tlyu committed:
update GD32F30x firmware library to 2.1.4

Vendor release 2.1.4 dated 2022-07-11. URLs:
https://www.gd32mcu.com/download/down/document_id/129/path_type/1
https://www.gd32mcu.com/data/documents/toolSoftware/GD32F30x_Firmware_Library_V2.1.4.rar

Notable USBD library improvements:

* Track control transfer state, avoiding some race conditions that
  caused looping ping-pong status transactions.

* Change some API types from 8-bit to 16-bit, avoiding truncation.

* Handle configuration descriptor sets longer than 255 bytes.

Signed-off-by: Taylor Yu <code@argon.blue>

tlyu committed:
undo incorrect vendor USBD change

USB-2.0 section 9.4.5 is fairly clear that ClearFeature(ENDPOINT_HALT)
resets the data toggles and clears the stall condition, regardless
of whether that feature had been previously explicitly set.

Instead, make the change that makes somewhat more sense: don't change
the status of a disabled endpoint.

Signed-off-by: Taylor Yu <code@argon.blue>

tlyu committed:
re-apply local USBD library fixes

Signed-off-by: Taylor Yu <code@argon.blue>

obra commented on system/GD32F30x_firmware/GD32F30x_standard_peripheral/Include/gd32f30x_ctc.h:
```diff
@@ -97,7 +97,6 @@ OF SUCH DAMAGE.
 #define CTL1_REFSEL(regval)                              (BITS(28,29) & ((uint32_t)(regval) << 28))
 #define CTC_REFSOURCE_GPIO                               CTL1_REFSEL(0)               /*!< GPIO is selected */
 #define CTC_REFSOURCE_LXTAL                              CTL1_REFSEL(1)               /*!< LXTAL is selected */
-#define CTC_REFSOURCE_USB_SOF                            CTL1_REFSEL(2)               /*!< USBD_SOF or USBFS_SOF is selected */
```
Would it be possible for us to locally revert the change that breaks the SOF CTC source?

tlyu replied:
Yeah, we could do that. Do you want that added to this PR or a separate one?

I know it appears to work, at least on the one piece of hardware that I have access to. Unfortunately, we still don't know why it was silently removed. We might not want to use it if it causes some non-obvious catastrophic behavior some of the time.

obra commented:
I have devices in the field that need it, but wouldn't depend on it for
anything new.

I'd rather it go in in the same PR, but in a standalone commit.

On Fri, Feb 24, 2023 at 1:20 PM Taylor Yu ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> system/GD32F30x_firmware/GD32F30x_standard_peripheral/Include/gd32f30x_ctc.h
> <https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/30#discussion_r1117666837>
> :
>
> > @@ -97,7 +97,6 @@ OF SUCH DAMAGE.
>  #define CTL1_REFSEL(regval)                              (BITS(28,29) & ((uint32_t)(regval) << 28))
>  #define CTC_REFSOURCE_GPIO                               CTL1_REFSEL(0)               /*!< GPIO is selected */
>  #define CTC_REFSOURCE_LXTAL                              CTL1_REFSEL(1)               /*!< LXTAL is selected */
> -#define CTC_REFSOURCE_USB_SOF                            CTL1_REFSEL(2)               /*!< USBD_SOF or USBFS_SOF is selected */
>
> Yeah, we could do that. Do you want that added to this PR or a separate
> one?
>
> I know it appears to work, at least on the one piece of hardware that I
> have access to. Unfortunately, we still don't know why it was silently
> removed. We might not want to use it if it causes some non-obvious
> catastrophic behavior some of the time.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/30#discussion_r1117666837>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2HCU3MTCPMRTERKNTTWZEQZRANCNFSM6AAAAAAVHLKPSY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>

tlyu committed:
restore CTC_REFSOURCE_USB_SOF

The 2.1.4 vendor firmware silently removed the USB SOF sync for the CTC
peripheral, without explanation. It appears to work anyway, on some
hardware.

Signed-off-by: Taylor Yu <code@argon.blue>
